### PR TITLE
Refactor function get_model

### PR DESCRIPTION
**Pull request by federicotdn**
**Proposed changes**:

This PR modifies the behavior of `get_model`, so that is raises an exception when it encounters an error instead of returning `None`. This has the following advantages:
- The return type of the function is now `UnpackedModelPath` (was: `Optional[UnpackedModelPath]`), which means that the function can now be used as a context manager without having to check for `None`.
- Different exceptions can be raised depending on the encountered error.

Other changes:
- Add class `RasaException` as a base class for all Rasa exceptions. Make `RasaCoreException` inherit from it. 
- A model was being loaded and then deleted, but not used, in rasa/run.py (function `run`).

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

### CODE REVIEW

File: rasa/core/agent.py [hunk 1]
```diff
@@ -348,14 +349,12 @@ def load(
...
-
-        if (
-            model_path is None
-            or not os.path.exists(model_path)
-            or not os.path.isdir(model_path)
-        ):
+        try:
+            if not model_path:
+                raise ModelNotFound()
+            elif not os.path.isdir(model_path):
```
Comment (wochinge) [inline rasa/core/agent.py]:
```suggestion
            elif os.path.isfile(model_path):
```

File: tests/core/test_model.py [hunk 2]
```diff
@@ -58,6 +59,20 @@ def test_get_model_context_manager(trained_model):
     assert not os.path.exists(unpacked)
 
 
+def test_get_model_exception():
+    with pytest.raises(ModelNotFound):
```
Comment (wochinge) [inline tests/core/test_model.py]: use https://docs.pytest.org/en/latest/parametrize.html

File: rasa/model.py [hunk 3]
```diff
@@ -65,17 +67,19 @@ def get_model(model_path: Text = DEFAULT_MODELS_PATH) -> Optional[UnpackedModelP
...
+    if os.path.isdir(model_path):
         model_path = get_latest_model(model_path)
+    elif not model_path.endswith(".tar.gz"):
+        # `model_path` does not point to a directory or a .tar.gz
+        model_path = None
 
-    if model_path:
-        return unpack_model(model_path)
+    if not model_path:
+        raise ModelNotFound()
```
Comment (wochinge) [inline rasa/model.py]: can we add some information to the exception from which path it tried to load the model?

Comment (wochinge) [reply rasa/model.py]: (also to the other locations where we raise this exception)

Comment (federicotdn) [reply rasa/model.py]: Should I include just the path or an additional message?
```python
raise ModelNotFound(model_path)
raise ModelNotFound("Path '{}' does not point to a .tar.gz file.".format(model_path))
```

Comment (federicotdn) [reply rasa/model.py]: Went with the second option.

**Comment by federicotdn** [comment]
@wochinge the code in Agent.load currently looks like this:
```python
try:
    if not model_path:
        raise ModelNotFound("No path specified.")
    elif os.path.isfile(model_path):
        model_path = get_model(model_path)
except ModelNotFound:
    raise ValueError(...)
```

Maybe we could extend `get_model` a bit so that it accepts paths containing `nlu` + `core` directories? This way we could avoid doing the extra check for `model_path` here. In this case, `get_model` would return a context manager that does nothing on exit, so that it doesn't delete the user's model.

Agent.load would then look like this:
```python
try:
    model_path = get_model(model_path)
except ModelNotFound:
    raise ValueError(...)
```

File: rasa/model.py [hunk 4]
```diff
@@ -66,16 +68,22 @@ def get_model(model_path: Text = DEFAULT_MODELS_PATH) -> Optional[UnpackedModelP
...
-        model_path = get_latest_model(model_path)
+        raise ModelNotFound("No file or directory at '{}'.".format(model_path))
 
-    if model_path:
-        return unpack_model(model_path)
+    if os.path.isdir(model_path):
+        model_path = get_latest_model(model_path)
+        if not model_path:
+            raise ModelNotFound(
+                "Could not find any .tar.gz files in '{}'.".format(model_path)
```
Comment (wochinge) [inline rasa/model.py]:
```suggestion
                "Could not find any Rasa model files in '{}'.".format(model_path)
```

File: rasa/model.py [hunk 5]
```diff
@@ -66,16 +68,22 @@ def get_model(model_path: Text = DEFAULT_MODELS_PATH) -> Optional[UnpackedModelP
...
+            )
+    elif not model_path.endswith(".tar.gz"):
+        raise ModelNotFound(
+            "Path '{}' does not point to a .tar.gz file.".format(model_path)
```
Comment (wochinge) [inline rasa/model.py]:
```suggestion
            "Path '{}' does not point to a Rasa model file.".format(model_path)
```

File: rasa/run.py [hunk 6]
```diff
@@ -36,8 +36,6 @@ def run(
     import rasa.nlu.run
     from rasa.core.utils import AvailableEndpoints
```
Comment (wochinge) [inline rasa/run.py]: can you please also remove the unused imports?

File: rasa/run.py [hunk 7]
```diff
@@ -36,8 +36,6 @@ def run(
...
 
-    model_path = get_model(model)
```
Comment (wochinge) [inline rasa/run.py]: did we really unpack a model before in a random path and then remove this part, while unpacking it to a different path in the server? oh dear 😄

Comment (federicotdn) [reply rasa/run.py]: We were!

**Comment by federicotdn** [comment]
Fixed the author of first commit.

**Pull request merged**